lib/model/compose.dart

```diff
@@ -0,0 +1,100 @@
+import 'dart:math';
+
+//
+// Put functions for nontrivial message-content generation in this file.
+//
+// If it's complicated enough to need tests, it should go in here.
+//
+
+// https://spec.commonmark.org/0.30/#fenced-code-blocks
+final RegExp _openingBacktickFenceRegex = (() {
+  // Recognize a fence with "up to three spaces of indentation".
+  // Servers don't recognize fences that start with spaces, as of Server 7.0:
+  //   https://chat.zulip.org/#narrow/stream/6-frontend/topic/quote-and-reply.20fence.20length/near/1588273
+  // but that's a bug, since those fences are valid in the spec.
+  // Still, it's harmless to make our own fence longer even if the server
+  // wouldn't notice the internal fence that we're steering clear of,
+  // and if servers *do* start following the spec by noticing indented internal
+  // fences, then this client behavior will be nice.
+  const lineStart = r'^ {0,3}';
+
+  // The backticks, captured so we can see how many.
+  const backticks = r'(`{3,})';
+
+  // The "info string" plus (meaningless) leading or trailing spaces or tabs.
+  // It can't contain backticks.
+  const trailing = r'[^`]*$';
+  return RegExp(lineStart + backticks + trailing, multiLine: true);
+})();
+
+/// The shortest backtick fence that's longer than any in [content].
+///
+/// Expressed as a number of backticks.
+///
+/// Use this for quote-and-reply or anything else that requires wrapping
+/// Markdown in a backtick fence.
+///
+/// See the CommonMark spec, which Zulip servers should but don't always follow:
+///   https://spec.commonmark.org/0.30/#fenced-code-blocks
+int getUnusedBacktickFenceLength(String content) {
+  final matches = _openingBacktickFenceRegex.allMatches(content);
+  int result = 3;
+  for (final match in matches) {
+    result = max(result, match[1]!.length + 1);
+  }
+  return result;
+}
+
+/// Wrap Markdown [content] with opening and closing backtick fences.
+///
+/// For example, for this Markdown:
+///
+/// ```javascript
+/// console.log('Hello world!');
+/// ```
```

Comment:
nit:
```suggestion
/// For example, for this Markdown:
///
///     ```javascript
///     console.log('Hello world!');
///     ```
```

and similarly below, so that this dartdoc renders properly, e.g. in the IDE at a call site